vignettes/simplify_function.Rmd

```diff
@@ -1,28 +1,121 @@
+Estimator's `predict()` and `evaluate()` provide an easy and efficient way of making predictions and evaluations on new data. However, the prediction and evaluation results are pretty verbose and usually in forms of nested lists, which isn't user-friendly. By default, the results produced from canned estimators will be simplified using a default simplify function for both `predict()` and `evaluate()`. You can also write a custom simplify function to them. Custom estimators do not simplify results by default unless you provide a custom simplify function. In this tutorial, we'll illustrate how the simplified results look like and how you can write your own simplify function.
 
-When you click the **Knit** button a document will be generated that includes both content as well as the output of any embedded R code chunks within the document. You can embed an R code chunk like this:
+## Simplify Prediction Results
 
-```{r cars}
-summary(cars)
+First of all, let's define our feature columns, input function, and a basic linear classifier:
+
+```{r}
+library(tfestimators)
+
+cols <- feature_columns( 
+  column_numeric("disp", "cyl")
+)
+
+mtcars_input_fn <- function(data) {
+  input_fn(data, 
+           features = c("disp", "cyl"), 
+           response = "vs")
+}
+
+indices <- sample(1:nrow(mtcars), size = 0.80 * nrow(mtcars))
+train_data <- mtcars[indices, ]
+test_data  <- mtcars[-indices, ]
+
+model <- linear_classifier(feature_columns = cols)
+```
+
+and we train the model using the input function constructed from training data:
+
+```{r}
+model %>% train(mtcars_input_fn(train_data))
+```
+
+Here's how the prediction output looks like without applying the simplifier on the results:
+
+```{r}
+model %>% predict(mtcars_input_fn(mtcars[1, ]), simplify = FALSE)
+```
+
+Note that here we only evaluate the first row of `mtcars` data since the results will be in forms of long nested lists that may not fit in this page.
+
+
+### Default Simplifier
+
+If `simplify = TRUE`, a default simplify function for predictions will be used to flatten the results so they are more appearing and concise like the following:
+
+```{r}
+model %>% predict(mtcars_input_fn(test_data), simplify = TRUE)
+```
+
+Note that this time we use the full test dataset since no matter how large the prediction results are, we can still see the results in a concise [tibble object](http://tibble.tidyverse.org/index.html). 
+
+### Custom Simplifier for Prediction Result
+
+You can also write your own custom simplify function to clean up the prediction results. For example, below we define a custom function that obtains the probabilities from the list of prediction results with keys being the prediction keys pre-defined for this canned estimator. In this case, `linear_classifier` has logits, probabilities, prediction classes, logistics, etc. pre-defined. We can see a full list of available predictions keys by calling `prediction_keys()`. 
```

Comment:
"clean up prediction results"

Reply:
I thought we were going to move away from the enum style "prediction_keys" and instead just use simple character vectors as is customary elsewhere in R?